iree/compiler/Dialect/Stream/Transforms/ScheduleAllocation.cpp

```diff
@@ -0,0 +1,1440 @@
+      continue;
+    }
+
+    locs.push_back(value.getLoc());
+    transientValues.push_back(value);
+    lifetimeIntervals.push_back(valueInterval.start);
+    lifetimeIntervals.push_back(valueInterval.end);
+
+    // Compute the allocation size for the value.
+    auto allocationSize = IREE::Util::SizeAwareTypeInterface::findSizeValue(
+        value, executeOp->getBlock(), Block::iterator(executeOp));
+    assert(allocationSize && "must have computable size");
+    dynamicSliceSizes.push_back(allocationSize);
+
+    // Mark as covered so we don't allocate it again.
+    scope.forEachResourceAlias(
+        value, [&](Value alias) { coveredValues.insert(alias); });
+  }
+  if (transientValues.empty()) {
+    // No transients required.
+    return llvm::None;
+  }
+
+  // Insert the stream.resource.pack op to compute the slices and total size of
+  // the transient allocation required.
+  auto fusedLoc = externalBuilder.getFusedLoc(locs);
+  auto indexType = externalBuilder.getIndexType();
+  SmallVector<Type> packedOffsetTypes(dynamicSliceSizes.size(), indexType);
+  auto packOp = externalBuilder.create<IREE::Stream::ResourcePackOp>(
+      fusedLoc, indexType, packedOffsetTypes,
+      /*offset=*/nullptr, externalBuilder.getIndexArrayAttr(lifetimeIntervals),
+      dynamicSliceSizes, executeOp.affinityAttr());
+
+  // Allocate the transient storage for the entire packed slab.
+  auto transientType = externalBuilder.getType<IREE::Stream::ResourceType>(
+      IREE::Stream::Lifetime::Transient);
+  auto timepointType = externalBuilder.getType<IREE::Stream::TimepointType>();
+  auto allocaOp = externalBuilder.create<IREE::Stream::ResourceAllocaOp>(
+      fusedLoc, transientType, timepointType, packOp.total_length(),
+      executeOp.affinityAttr());
+  TransientAllocation allocation;
+  allocation.awaitTimepoint = allocaOp.result_timepoint();
+  allocation.reservation = allocaOp.result();
+  allocation.reservationSize = allocaOp.storage_size();
+  allocation.capturedArg = executeOp.body().front().addArgument(
+      allocation.reservation.getType(), allocation.reservation.getLoc());
+
+  // Map values to their ranges within the slab.
+  for (size_t i = 0; i < transientValues.size(); ++i) {
+    auto value = transientValues[i];
+    auto offset = packOp.packed_offsets()[i];
+    auto length = packOp.dynamic_slice_sizes()[i];
+    auto resourceRange = ResourceRange(
+        allocation.capturedArg, allocation.reservationSize, offset, length);
+    scope.mapResourceRange(value, resourceRange);
+  }
+
+  return allocation;
+}
+
+//===----------------------------------------------------------------------===//
+// Constant allocation
+//===----------------------------------------------------------------------===//
+
+struct ConstantReservation {
+  IREE::Stream::AsyncConstantOp constantOp;
+  Value resource;
+  Value resourceSize;
+  // NOTE: may be nullptr if the constant is was not used within the region.
+  BlockArgument capturedArg;
```

Comment:
nit: grammar ("is not" or "was not", not "is was not")

```suggestion
  // NOTE: may be nullptr if the constant is not used within the region.
  BlockArgument capturedArg;
```